include/mata/utils/partition-relation-pair.hh

```diff
@@ -0,0 +1,1449 @@
+    
+    public:
+    
+        // constructors
+        HashedSquareMatrix(size_t maxRows, size_t initRows = 0);
+        
+        // implemented virtual functions        
+        inline void set(size_t i, size_t j, T value) override;
+        inline T get(size_t i, size_t j) const override;
+        inline void extend(T placeholder = T()) override;
+
+        // cloning
+        HashedSquareMatrix *clone(void) const 
+            { return new HashedSquareMatrix(*this); }
+        
+};
+
+/**
+* @brief creates a Hashed square matrix
+* @param maxRows capacity of the square matrix
+* @param initRows initial size of the square matrix
+*/
+template <typename T>
+HashedSquareMatrix<T>::HashedSquareMatrix
+    (size_t maxRows, size_t initRows)
+{
+    assert(initRows <= maxRows && 
+    "Initial size of the matrix cannot be bigger than the capacity");
+    
+    this->m_type = MatrixType::Hashed;
+    this->m_capacity = maxRows;
+    
+    // creating the initial size and filling the data cells with
+    // default values
+    for(size_t i = 0; i < initRows; ++i) {extend();}
+}
+
+/**
+* @brief assings a value to the Hashed square matrix
+* @param i row of the square matrix
+* @param j column of the square matrix
+* @param value value to be assigned to the square matrix data cell
+*/
+template <typename T>
+inline void HashedSquareMatrix<T>::set(size_t i, size_t j, T value)
+{
+    assert(i < this->m_size && "Nonexisting row cannot be accessed");
+    assert(j < this->m_size && "Nonexisting column cannot be accessed");
+    
+    // accessing the hashmap using row matrix traversal
+    data[i * this->m_capacity + j] = value;
+}
+
+/**
+* @brief returns a value of the Hashed square matrix
+* @param i row of the square matrix
+* @param j column of the square matrix
+* @return value found in the square matrix data cell
+*/
+template <typename T>
+inline T HashedSquareMatrix<T>::get(size_t i, size_t j) const
+{
+    assert(i < this->m_size && "Nonexisting row cannot be accessed");
+    assert(j < this->m_size && "Nonexisting column cannot be accessed");
+    
+    // accessing the hashmap using row matrix traversal
+    return data[i * this->m_capacity + j];
+}
+
+/**
+* @brief extends the Hashed square matrix by a new row and column
+* @param placeholder a value which will be assigned to all the new data cells
+*/
+template <typename T>
+inline void HashedSquareMatrix<T>::extend(T placeholder)
+{
+    assert(this->m_size < this->m_capacity 
+          && "Matrix cannot be extened anymore");
+
+    // creating a new row and column
+    for(size_t i = 0; i < this->m_size; ++i)
+    {
+        data[this->m_size * this->m_capacity + i] = placeholder;
+        data[i * this->m_capacity + this->m_size] = placeholder;
+    }
+    data[this->m_size * this->m_capacity + this->m_size] = placeholder;
+    
+    // increasing size
+    ++this->m_size;
+}
+
+/*************************************
+*
+*        ADDITIONAL FUNCTIONS
```

Comment:
@Adda0 I have originally decided not to inline these functions since they

- contain switch statement or several loops
- won't be used frequently 
    - the `operator<<` for `ExtendableSquareMatrix` was primarily designed for debugging purposes
    - the `is_reflexive`, `is_antisymetric` and `is_transitive` functions were primarily designed for `assertions` so any instance of `PartitionRelationPair` where the relation is not a proper partial order cannot be created. If I got it right, the `assertions` won't be ever invoked when compiling with `release`.
    - the `create` function was the only one which I considered for inlining but I did not do it

However, you have mentioned earlier that there is no need to inline any class method. The functions `is_reflexive`, `is_antisymetric` and `is_transitive` belong to the `ExtendableSquareMatrix` class. Could I have misunderstood something?

Reply:
Ah, sorry. The way the file is laid out, I thought that these additional functions are functions, not methods. If they were functions with definition in header file, they should be inlined (C++ inline is something different from C inline), otherwise there will be compilation errors during linking when including the file multiple times. See https://softwareengineering.stackexchange.com/a/339490.

Reply:
@Adda0 Thanks for the clarification! The explanation provided in the attached link is really nice. At the moment (after my recent commit):

- the function `operator<<` for the `ExtendableSquareMatrix` class is marked as `inline`
- the function `create` is marked as `inline`
- the function `operator<<` for the `Partition` class is not marked as `inline`. Although it is not a method, it is a friend function of the `Partition` class. If I got it right, even the `friend` functions declared in the class are automatically inlined
- the other provided functions are methods so they are not marked as `inline`

You are right, the placing of `ifReflexive, isAntisymetric` and `isTransivite` methods in the file did not make any sense. Initially, I had implemented them as ordinary functions. Then, I decided to rewrite them as class methods, but I did not move them under the corresponding class. I have fixed it in the recent commit.

Reply:
Looks good to me now.